### PR TITLE
Rewrite CIS1 examples to CIS2

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Update CIS1 token examples to CIS2.
Remove the uninteresting token examples.
Closing #106.

## Changes

- Update CIS1 NFT, fungible(wrapped CCD) and multi token examples to CIS2.
- Remove `erc721`, `cis1-single-nft` and `listing-cis1` examples.
- Replace `concordium-cis1` with `concordium-cis2` crate.
- Use LEB128 to encode token amounts

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/cis2-multi/src/lib.rs [hunk 1]
```diff
@@ -429,11 +429,11 @@ fn contract_transfer<S: HasStateApi>(
 
         // If the receiver is a contract we invoke it.
         if let Receiver::Contract(address, function) = to {
-            let parameter = OnReceivingCis1Params {
+            let parameter = OnReceivingCis2Params {
                 token_id,
                 amount,
                 from,
-                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS1-Multi")),
+                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS2-Multi")),
```
Comment (Bargsteen) [inline examples/cis2-multi/src/lib.rs:436]: Is this still needed?

File: examples/cis2-nft/src/lib.rs [hunk 2]
```diff
@@ -376,7 +376,7 @@ type TransferParameter = TransferParams<ContractTokenId>;
 /// - Any of the messages sent to contracts receiving a transfer choose to
```
Comment (Bargsteen) [inline examples/cis2-nft/src/lib.rs:376]: Consider rephrasing this doc-comment to remove "messages" and "actions" as that terminology belongs to v0 contracts.

File: examples/cis2-nft/src/lib.rs [hunk 3]
```diff
@@ -418,11 +418,11 @@ fn contract_transfer<S: HasStateApi>(
         // If the receiver is a contract, we add sending it a message to the list of
         // actions.
         if let Receiver::Contract(address, function) = to {
-            let parameter = OnReceivingCis1Params {
+            let parameter = OnReceivingCis2Params {
                 token_id,
                 amount,
                 from,
-                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS1-NFT")),
+                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS2-NFT")),
```
Comment (Bargsteen) [inline examples/cis2-nft/src/lib.rs:425]: Is it still necessary to include the contract name? Wasn't that just for the callback?

File: examples/cis2-wccd/src/lib.rs [hunk 4]
```diff
@@ -465,11 +465,11 @@ fn contract_transfer<S: HasStateApi>(
 
         // If the receiver is a contract, we invoke it.
         if let Receiver::Contract(address, function) = to {
-            let parameter = OnReceivingCis1Params {
+            let parameter = OnReceivingCis2Params {
                 token_id,
                 amount,
                 from,
-                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS1-Multi")),
+                contract_name: OwnedContractName::new_unchecked(String::from("init_CIS2-Multi")),
```
Comment (Bargsteen) [inline examples/cis2-wccd/src/lib.rs:472]: To be removed ;)

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Looks good in general. Needs a few tweaks as discussed 😊

File: examples/cis2-multi/src/lib.rs [hunk 5]
```diff
@@ -373,9 +373,8 @@ type TransferParameter = TransferParams<ContractTokenId>;
 
 /// Execute a list of token transfers, in the order of the list.
 ///
-/// Logs a `Transfer` event for each transfer in the list.
-/// Produces an action which sends a message to each contract which was the
-/// receiver of a transfer.
+/// Logs a `Transfer` event and invoke a receive hook function for every
```
Comment (abizjak) [inline examples/cis2-multi/src/lib.rs:376]:
```suggestion
/// Logs a `Transfer` event and invokes a receive hook function for every
```

**Review by abizjak — APPROVED** [review]

File: examples/cis2-wccd/src/lib.rs [hunk 6]
```diff
@@ -406,9 +399,8 @@ type TransferParameter = TransferParams<ContractTokenId>;
 
 /// Execute a list of token transfers, in the order of the list.
 ///
-/// Logs a `Transfer` event for each transfer in the list.
-/// Produces an action which sends a message to each contract which was the
-/// receiver of a transfer.
+/// Logs a `Transfer` event and invoke a receive hook function for every
```
Comment (Bargsteen) [inline examples/cis2-wccd/src/lib.rs:402]:
```suggestion
/// Logs a `Transfer` event and invokes a receive hook function for every
```

**Review by Bargsteen — APPROVED** [review]

File: concordium-cis2/src/lib.rs [hunk 7]
```diff
@@ -435,19 +454,476 @@ impl Deserial for TokenIdUnit {
...
+
+impl ops::DivAssign for TokenAmountU32 {
+    fn div_assign(&mut self, other: Self) { *self = *self / other; }
+}
+
+impl IsTokenAmount for TokenAmountU32 {}
+
+impl schema::SchemaType for TokenAmountU32 {
+    // TODO Fix the schema when supporting LEB128
+    fn get_type() -> schema::Type { schema::Type::Array(5, Box::new(schema::Type::U8)) }
```
Comment (abizjak) [inline concordium-cis2/src/lib.rs:688]: Why is this a fixed array?

**Review by abizjak — COMMENTED** [review]
I think overall it looks fine.

It would be useful to reduce the boilerplate a bit. Perhaps by using `derive_more` and adding a couple of macros to generate the implementations for serialization of token amounts (just normal macros, not procedural ones).

File: concordium-cis2/src/lib.rs [hunk 8]
```diff
@@ -435,19 +454,476 @@ impl Deserial for TokenIdUnit {
...
+        let mut result = 0u128;
+        for i in 0..37 {
+            let byte = source.read_u8()?;
+            let value_byte = (byte & 0b0111_1111) as u128;
+            result = result.checked_add(value_byte << (i * 7)).ok_or(ParseError {})?;
+
+            if byte & 0b1000_0000 == 0 {
+                return Ok(TokenAmountU128(result));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (limemloh) [reply concordium-cis2/src/lib.rs:536]: I can add an issue for looking into this and/or add a comment in the code

File: concordium-cis2/src/lib.rs [hunk 9]
```diff
@@ -435,19 +454,476 @@ impl Deserial for TokenIdUnit {
...
+            let byte = source.read_u8()?;
+            let value_byte = (byte & 0b0111_1111) as u64;
+            result = result.checked_add(value_byte << (i * 7)).ok_or(ParseError {})?;
+
+            if byte & 0b1000_0000 == 0 {
+                return Ok(TokenAmountU64(result));
+            }
+        }
+        Err(ParseError {})
+    }
```
[the remark this answers is not in the document]
Comment (limemloh) [reply concordium-cis2/src/lib.rs]: I know it was not exactly what you ment, but I have made macro with all the token amounts and thier implementations of basic operations.

File: concordium-cis2/src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-cis2/src/lib.rs]: Add an issue, linking to this example of Wasm to make sure it is not forgotten.

File: concordium-cis2/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply concordium-cis2/src/lib.rs]: Looks good.

File: concordium-cis2/src/lib.rs (hunk 7 above)
Comment (abizjak) [reply concordium-cis2/src/lib.rs]: Ok. Hopefully resolved very soon.

**Review by abizjak — APPROVED** [review]

File: concordium-cis2/src/lib.rs [hunk 10]
```diff
@@ -774,37 +903,48 @@ impl<T: IsTokenId> schema::SchemaType for Transfer<T> {
...
-#[derive(Debug, Serialize, SchemaType)]
+#[derive(Debug, Serialize)]
 pub enum OperatorUpdate {
     /// Remove the operator.
     Remove,
     /// Add an address as an operator.
     Add,
 }
 
+impl schema::SchemaType for OperatorUpdate {
```
Comment (Bargsteen) [inline concordium-cis2/src/lib.rs:935]: Why was this implemented manually? 😊

Comment (Bargsteen) [reply concordium-cis2/src/lib.rs:935]: And all the ones below.

**Review by Bargsteen — APPROVED** [review]
LGTM 👍

**Pull request merged**